packages/gatsby/src/utils/jobs-manager.js

```diff
@@ -0,0 +1,238 @@
+ * @param {Job} job
+ * @return Promise<T>
+ */
+const runLocalWorker = async (workerFn, job) => {
+  await fs.ensureDir(job.outputDir)
+
+  return new Promise((resolve, reject) => {
+    // execute worker nextTick
+    // TODO should we think about threading/queueing here?
+    process.nextTick(() => {
+      try {
+        resolve(
+          workerFn({
+            inputPaths: job.inputPaths,
+            outputDir: job.outputDir,
+            args: job.args,
+          })
+        )
+      } catch (err) {
+        reject(err)
+      }
+    })
+  })
+}
+
+/**
+ *
+ * @param {AugmentedJob} job
+ */
+const runJob = ({ plugin, ...job }) => {
+  const worker = require(`${plugin.name}/worker.js`)
+  if (!worker[job.name]) {
+    throw new Error(`No worker function found for ${job.name}`)
+  }
+
+  return runLocalWorker(worker[job.name], job)
+}
+
+const handleJobEnded = () => {
+  if (--activeJobs === 0) {
+    activityForJobs.end()
+    activityForJobs = null
+  }
+}
+
+/**
+ * Create an internal job object
+ *
+ * @param {Job|AugmentedJob} job
+ * @param {{name: string, version: string}} plugin
+ * @param {string} rootDir
+ * @return {AugmentedJob}
+ */
+exports.createInternalJob = (job, plugin, rootDir) => {
+  // It looks like we already have an augmented job so we shouldn't redo this work
+  if (job.id && job.contentDigest) {
+    console.log(`cached`)
+    return job
+  }
+
+  const { name, inputPaths, outputDir, args } = job
+
+  // TODO see if we can make this async, filehashing might be expensive to wait for
+  // currently this needs to be sync as we could miss jobs to have been scheduled and
+  // are still processing their hashes
+  const inputPathsWithContentDigest = inputPaths.map(path => {
+    return {
+      path: convertPathsToRelative(path, rootDir),
+      contentDigest: createFileHash(path),
+    }
+  })
+
+  const augmentedJob = {
+    id: uuid(),
+    name,
+    inputPaths: inputPathsWithContentDigest,
+    outputDir: convertPathsToRelative(outputDir, rootDir),
+    args,
+    plugin,
+  }
+
+  augmentedJob.contentDigest = createContentDigest({
+    name: job.name,
+    inputPaths: augmentedJob.inputPaths.map(
+      inputPath => inputPath.contentDigest
+    ),
+    outputDir: augmentedJob.outputDir,
+    args: augmentedJob.args,
+    plugin: augmentedJob.plugin,
+  })
+
+  return augmentedJob
+}
+
+/**
+ * Creates a job
+ *
+ * @param {AugmentedJob} job
+ * @return {Promise<unknown>}
+ */
+exports.enqueueJob = async job => {
+  // When we already have a job that's executing, return the same promise.
+  if (jobsInProcess.has(job.contentDigest)) {
+    return jobsInProcess.get(job.contentDigest).deferred.promise
+  }
+
+  // Bump active jobs
+  activeJobs++
+  if (!activityForJobs) {
+    activityForJobs = reporter.phantomActivity(`Running jobs`)
+    activityForJobs.start()
+  }
+
+  const deferred = pDefer()
+  jobsInProcess.set(job.contentDigest, {
+    id: job.id,
+    deferred,
+  })
+
+  try {
```

Comment:
Since `runJob` already returns a promise, couldn't you use that instead of creating a new `pDefer()`?

```js
try {
  // The async code will first run in sync up to the first `await` or `return` but always returns a promise, so this is fine...
  const promise = runJob(job)

  // Potential memory leak at scale, but caching is nice too
  jobsInProcess.set(job.contentDigest, {
    id: job.id,
    deferred: promise, // Seems this is only a cache, not actually used to `.resolve()` or `.reject()`, otherwise it won't work
  })

  await promise
} finally {
    handleJobEnded()
}
```

Reply:
Admittedly, the catch is that up to the first yielding point in `runJob`, the `jobsInProcess` map will not contain the current running job. It's a little brittle to assume a job won't start itself in sync, but technically that could happen.

Reply:
Yeah actually that won't do. This may be too big of a footgun for various reasons. Let's keep the pdefer.